README.md
## SwiftUI-Introduction-View

This is a simple IOS SwiftUI application that demonstrates how to integrate custom introduction screen with customizable features in your swiftUI projects.

## Table of Contents 

- [Demo UI](#demo)
- [Features](#features)
- [Installation](#installation)
- [Usage](#usage)
- [Customization](#customization)
- [Contributing](#contributing)
- [Need Help / Support?](#need-help)
- [Collection of Components](#collection-of-Components)
- [License](#license)
- [Acknowledgements](#acknowledgements)
- [Keywords](#keywords)

# Demo UI 📱


<img src="./introductionview.gif" alt="Demo Preview">

# Features 🥳

App has main feature as:

* Page indicator visibility and style can be customized."
* Easily swipe page content left and right.
* Page UI elements are fully configurable.
* Easily integrated by using model data.
* Navigation buttons can be easily customized.
* Customize the font and colors of page elements.
* Fully customizable.
* No third-party dependencies.
* Supports the latest iOS 17.0.

# Usage 👩🏾‍🔬
To use Introduction screen features you need to drag and drop the `UtilHelper` folder in your project to integrate introduction view feature in your project, follow these steps:

Open your `ContentView` file to add the following code snippet like:

```swift

VStack{
    CustomIntroductionView(config: config())
}

```

Customise your own Introduction Screen by creating your own `config()` method and pass it as parameter into `CustomIntroductionView()`. Eg. `func config() -> CustomIntroConfig`

# Customization 

These are the customizable properties

```swift

func config() -> CustomIntroConfig {
    var config = CustomIntroConfig(listIntroData: listIntroData, isShowPageIndicator: true)
    config.pageIndicatorAlignment = .bottom
    config.pageIndicatorColor = .indigo
    config.titleFontSize = 24
    config.titleFontColor = .indigo
    config.quoteFontColor = .gray
    config.quoteFontWeight = .thin
    config.descriptionFontSize = 19
    config.descriptionFontColor = .black
    config.imageCornerRadius = 15
    config.isButtonPageIndicatorOverlay = true
    config.buttonAction = {print("started")}
    config.isShowButton = true
    return config
}

```

### Required and Common Props

| Props                       | Type           | Description                                                          | Required | Default |
| :-------------------------- | :------------- | :------------------------------------------------------------------- | :------- | :------ |
| listIntroData               | Array          | array of data for the Introduction details to be render              | required |
| isShowPageIndicator                | Bool            | to show pageIndicator on view                                               | required |
| isPageIndicatorCircle              | Bool           | to show pageIndicator in circle                                             |          | false   |
| isButtonPageIndicatorOverlay       | Bool           | to overlay button/pageIndicator on view                                     |          | false   |
| isShowButton                | Bool           | to show button on view                                               |          | false   |
| introViewBackgroundColor    | Color          | used to set background colour of introduction view                   |          | .white  |
| buttonPageIndicatorBackgroundColor | Color          | used to set background colour of button and pageIndicator                   |          | .white  |


### Default value of UI element properties 

| Props                      | Type        | Description                    | Default   |
| :------------------------- | :---------- | :----------------------------- | :--------------------- |
| imageWidth                 | CGFloat     | set image width                |   330                  |
| imageHeight                | CGFloat     | set image height               |   15      |
| imageCornerRadius          | CGFloat     | set image corner radius        |   0       |
| titleFontSize              | CGFloat     | set custom title font size     |   20      |
| titleFontColor             | Color       | set title font color           |  .black   |
| titleFontWeight            | Font.Weight | set title font weight          |  .bold    |
| quoteFontSize              | CGFloat     | set custom quote font size     |   15      |
| quoteFontColor             | Color       | set quote font color           |  .black   |
| quoteFontWeight            | Font.Weight | set quote font weight          |  .regular |
| descriptionFontSize        | CGFloat     | set description font size      |   15      |
| descriptionFontColor       | Color       | set description font color     |  .black   |
| descriptionFontWeight      | Font.Weight | set description font weight    |  .regular |
| pageIndicatorAlignment.           | Alignment   | set the pageIndicator alignment       |  .bottom  |
| pageIndicatorColor                | Color       | set the pageIndicator color           |  .black  |
| buttonTextSize             | CGFloat     | set button text size           |  .15            |
| buttonTextWeight           | Font.Weight | set button text weight         |  .regular |
| buttonTextTitle            | String      | set button title               |  "Get Started" |
| buttonTextTitleColor       | Color       | set button title color         |  .white       |
| buttonTextBackgroundColor  | Color       | set button background color    |  .black |
| buttonCornerRadius         | CGFloat     | set button corner radius       |  15 |
| buttonPadding              | Edge.Set    | set button padding             |  .all             |
| buttonAction               | () -> Void  | set button action              |  {print("button Tapped")} |   //22
| buttonTextTitle            | String      | set button title               |  "Get Started" |
| buttonTextTitle            | String      | set button title               |  "Get Started" |



# Installation 💾

The instructions on how to install or run this demo:

* Clone the repository.
* Open the project in Xcode.
* Build and run the project.

# Contributing

- Created something awesome, made this code better, added some functionality, or whatever (this is the hardest part).
- [Fork it](http://help.github.com/forking/).
- Create new branch to contribute your changes.
- Commit all your changes to your branch.
- Submit a [pull request](http://help.github.com/pull-requests/).


# Need Help? 

We also provide a free, basic support for all users who want to use this coordinator master demo in project. In case you want to customize this demo input to suit your development needs, then feel free to contact our [iOS Developers](https://www.weblineindia.com/hire-ios-app-developers.html).

# Collection of Components

We have built many other components and free resources for software development in various programming languages. Kindly click here to view our [Free Resources for Software Development.](https://www.weblineindia.com/communities.html)

# Changelog

Detailed changes for each release are documented in [CHANGELOG](./CHANGELOG).

# License 📃

This project is licensed under the MIT License. Feel free to use, modify, and distribute the code as permitted by the license.

# Acknowledgements

This app was created using SwiftUI and leverages various libraries and resources. We would like to acknowledge and express our gratitude to the following:
* SwiftUI: Apple's declarative framework for building user interfaces.
* Xcode: Integrated development environment (IDE) provided by Apple.
* Open-source libraries: Various open-source libraries and frameworks used in the project.
* Community contributors: Contributors who have provided support, bug fixes, and enhancements.

# Keywords

Custom Intro screen swiftUI, Tutorial view component, onboarding screen SwiftUI, Custom info screen, Custom introduction view, SwiftUI5, iOS 17, WeblineIndia




CHANGELOG
1.0 (December 19, 2023)
Initial release - Sample source code of Introduction view.


Introduction_view/IntroductionMainView.swift
//
//  IntroductionView.swift
//  Introduction_view
//
//  Created by wli's on 12/12/23.
//

import SwiftUI

struct IntroductionMainView: View {
    
    var body: some View {
        VStack{
            
            CustomIntroductionView(config: config())
        }
    }
    
    func config() -> CustomIntroConfig {
        var config = CustomIntroConfig(listIntroData: listIntroData, isShowPageIndicator: true)
        config.pageIndicatorAlignment = .bottom
        config.pageIndicatorColor = .indigo
        config.titleFontSize = 24
        config.titleFontColor = .indigo
        config.quoteFontColor = .gray
        config.quoteFontWeight = .thin
        config.descriptionFontSize = 19
        config.descriptionFontColor = .black
        config.imageCornerRadius = 15
        config.isButtonPageIndicatorOverlay = true
        config.buttonAction = {print("started")}
        config.isShowButton = true
        return config
    }
}



Introduction_view/UtilHelper/IntroPagesModel.swift
//
//  View.swift
//  Introduction_view
//
//  Created by IOSUSER on 12/12/23.
//

import SwiftUI

struct IntroModel:Identifiable {
    var id = UUID()
    var image: String?
    var title: String?
    var quote: String?
    var description: String?
    var tag: Int
    
    static var nextTag = 0
      
      init(image: String?, title: String?, quote: String?, description: String?) {
          self.image = image
          self.title = title
          self.quote = quote
          self.description = description
          self.tag = IntroModel.nextTag
          IntroModel.nextTag += 1
      }
}



var listIntroData = [
        IntroModel(image: "solarsystem", title: "Solar System", quote: "When you look at the stars and the galaxy, you feel that you are not just from any particular piece of land, but from the solar system.", description: "The Solar System consists of the Sun Moon and Planets. It also consists of comets, meteoroids and asteroids. The Sun is the largest member of the Solar System. In order of distance from the Sun, the planets are Mercury, Venus, Earth, Mars, Jupiter, Saturn, Uranus, Neptune and Pluto; the dwarf planet. The Sun is at the centre of the Solar System and the planets, asteroids, comets and meteoroids revolve around it."),
        IntroModel(image: "sun", title: "Sun", quote: "The Sun never repents of the good he does, nor does he ever demand a recompense.", description: "The sun is a huge ball of gases. It has a diameter of 1,392,000 km. It is so huge that it can hold millions of planets inside it. The Sun is mainly made up of hydrogen and helium gas. The surface of the Sun is known as the photosphere. The photosphere is surrounded by a thin layer of gas known as the chromospheres. Without the Sun, there would be no life on Earth. There would be no plants, no animals and no human beings. As, all the living things on Earth get their energy from the Sun for their survival."),
        
        IntroModel(image: "earthfinal", title: "Earth", quote: "Those who contemplate the beauty of the earth find reserves of strength that will endure as long as life lasts.", description: "Earth is the third planet in our solar system. It is in the shape which is close to sphere (spheroid). Earth is called the “blue planet” because it looks blue in colour when it is seen from outer space. It looks blue because there is water on this planet. It is the only planet that has life on it. Life is possible on Earth because of water and because of all the gases which are important for life. Earth is in perfect distance from Sun and so there is a mild climate here. Other planets are too hot or too cold and life is impossible there."),
        
        IntroModel(image: "moon", title: "Moon", quote: "Do not swear by the moon, for she changes constantly. then your love would also change.", description: "The Moon is a barren, rocky world without air and water. It has dark lava plain on its surface. The Moon is filled wit craters. It has no light of its own. It gets its light from the Sun. The Moo keeps changing its shape as it moves round the Earth. It spins on its axis in 27.3 days stars were named after the Edwin Aldrin were the first ones to set their foot on the Moon on 21 July 1969 They reached the Moon in their space craft named Apollo II."),
        
        
    ]


Introduction_view/UtilHelper/CustomIntroductionView.swift
//
//  CustomIntroductionView.swift
//  Introduction_view
//
//  Created by IOSUSER on 12/12/23.
//

import SwiftUI

struct CustomIntroductionView: View {
    @State private var selectedTab = 0
    var config: CustomIntroConfig
    var body: some View {
        ZStack {
            
            VStack {
                TabView(selection: $selectedTab) {
                    ForEach(config.listIntroData.indices, id: \.self) { index in
                        let data = config.listIntroData[index]
                        ScrollView(showsIndicators: false){
                            Spacer(minLength: 30)
                            VStack(spacing: 20) {
                                if let image = data.image {
                                    Image(image)
                                        .resizable()
                                        .scaledToFill()
                                        .frame(width: config.imageWidth, height: config.imageHeight)
                                    
                                        .cornerRadius(config.imageCornerRadius ?? 0)
                                }
                                
                                if let title = data.title {
                                    Text(title)
                                        .font(.system(size: config.titleFontSize ?? 20))
                                    
                                        .foregroundColor(config.titleFontColor)
                                        .fontWeight(config.titleFontWeight)
                                }
                                
                                if let quote = data.quote {
                                    Text("\"\(quote)\"")
                                        .font(.system(size: config.quoteFontSize ?? 15))
                                        .foregroundColor(config.quoteFontColor)
                                        .fontWeight(config.quoteFontWeight)
                                }
                                
                                if let description = data.description {
                                    Text(description)
                                        .font(.system(size: config.descriptionFontSize ?? 15))
                                        .foregroundColor(config.descriptionFontColor)
                                        .fontWeight(config.descriptionFontWeight)
                                        .multilineTextAlignment(.leading)
                                        
                                }
                            }
                            .padding(.horizontal, 10)
                            
                            Spacer(minLength: 100)

                        }
                        .padding(.bottom)
                        .tag(index)
                        
                    }
                    
                }
                
            }
        }
        .background(config.introViewBackgroundColor)
        .tabViewStyle(.page(indexDisplayMode: .never))
        .overlay(
            VStack(alignment: .center, spacing: 30){
                
                HStack{
                    if config.listIntroData.count > 1{
                        ForEach(config.listIntroData.indices, id: \.self) { index in
                            let isSelected = index == selectedTab
                            if config.isShowPageIndicator{
                                if config.isPageIndicatorCircle ?? false{
                                    Circle()
                                        .frame(width: isSelected ? 13 : 8, height: isSelected ? 13 : 8)
                                        .foregroundColor(config.pageIndicatorColor)
                                        .shadow(color: .black, radius: 1)
                                }

                                else{
                                    withAnimation(.easeInOut(duration: 1.5)) {
                                        Capsule()
                                            .frame(width: isSelected ? 40 : 10, height: 10)
                                            .foregroundColor(config.pageIndicatorColor)
                                            .shadow(color: .black, radius: 1)

                                    }
                                   
                                }
                            }
                        }
                    }
                    
                }
                if config.isShowButton ?? false{
                    HStack{
                        Button(action: {
                            (config.buttonAction ?? { print("Button Tapped") })()
                        }) {
                            Text(config.buttonTextTitle ?? "Get Started")
                                .font(.system(size: config.buttonTextSize ?? 15))
                                .fontWeight(config.buttonTextWeight)
                                .foregroundColor(config.buttonTextTitleColor)
                                .padding(config.buttonPadding ?? .all)
                                .background(config.buttonTextBackgroundColor)
                                .cornerRadius(config.buttonCornerRadius ?? 15)
                        }
                    }
                }
                
            }
                .frame(minWidth: 0, maxWidth: .infinity)
                .padding(.bottom, 5)
                .padding(.top, 20)
                .animation(.spring(), value: selectedTab)
                .background(config.isButtonPageIndicatorOverlay ?? false ? Color.clear : config.buttonPageIndicatorBackgroundColor  )
            , alignment: config.pageIndicatorAlignment ?? .bottom)
    }
}


Introduction_view/UtilHelper/CustomIntroConfig.swift
//
//  CustomIntroConfig.swift
//  Introduction_view
//
//  Created by IOSUSER on 13/12/23.
//

import SwiftUI

struct CustomIntroConfig {
    var listIntroData: [IntroModel]
    var isShowPageIndicator: Bool
    var isPageIndicatorCircle: Bool? = false
    var isButtonPageIndicatorOverlay: Bool? = false
    var isShowButton: Bool? = false
    var pageIndicatorAlignment: Alignment? = .bottom
    var pageIndicatorColor: Color? = .black
    var descriptionFontSize: CGFloat? = 15
    var descriptionFontColor: Color? = .black
    var descriptionFontWeight: Font.Weight? = .regular
    var titleFontSize: CGFloat? = 20
    var titleFontColor: Color? = .black
    var titleFontWeight: Font.Weight? = .bold
    var quoteFontSize:  CGFloat? = 15
    var quoteFontColor: Color? = .black
    var quoteFontWeight: Font.Weight? = .regular
    var imageWidth: CGFloat? = 330
    var imageHeight: CGFloat? = 250
    var imageCornerRadius: CGFloat? = 0
    var buttonTextSize: CGFloat? = 15
    var buttonTextWeight: Font.Weight? = .regular
    var buttonTextTitle: String? = "Get Started"
    var buttonTextTitleColor: Color? = .white
    var buttonTextBackgroundColor: Color? = .black
    var buttonCornerRadius: CGFloat? = 15
    var buttonPageIndicatorBackgroundColor: Color? = .white
    var introViewBackgroundColor: Color? = .white
    var buttonPadding: Edge.Set? = .all
    var buttonAction: (() -> Void)? = {print("button Tapped")}
    
}
